### PR TITLE
Add retry support to GpuOutOfCoreSortIterator.mergeSortEnoughToOutput 

### PR DESCRIPTION
**Pull request by firestarman**
closes https://github.com/NVIDIA/spark-rapids/issues/8351

This PR adds in retry support for more operations in `GpuOutOfCoreSortIterator`, including computing the split offset and bringing the data back to GPU to remove the projected columns.

Besides, to keep being eager to close the input batches in the `mergeSortAndClose` function (introduced by https://github.com/NVIDIA/spark-rapids/pull/6931), instead of retrying the call to the whole `mergeSortAndClose` function, we retry the operations inside it, including bringing the data back to GPU, concatenating tables, sort the concatenated table and merging the input tables.

It also covers a small followup change in `GpuColumnToRowExec` for PR #9088.

### CODE REVIEW

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala [hunk 1]
```diff
@@ -261,20 +261,27 @@ class GpuSorter(
               // It is slower, but it works
               val concatenated = closeOnExcept(tablesToMerge) { _ =>
                 while (spillableBatches.nonEmpty) {
-                  withResource(spillableBatches.pop()) { sb =>
+                  // We retry work with a single spillable batch, then it can be closed
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala:264]: So do you plan on having a follow on issue to fix this?

If we need to make tablesToMerge Spillable we should do that.

Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala]: Sorry for cunfusion, and it is no longer an issue. Please ignore it.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
@revans2 Could you review it again ?
@abellina Your review would be also appreciated, especially for the SortUtils part.

**Review by revans2 — APPROVED** [review]

**Pull request merged**